consensus/ethash/ethash.go

```diff
@@ -226,31 +226,34 @@ func (lru *lru) get(epoch uint64, epochLength uint64, ecip1099FBlock *uint64) (i
 	lru.mu.Lock()
 	defer lru.mu.Unlock()
 
+	cacheKey := fmt.Sprintf("%d-%d", epoch, epochLength)
 	// Get or create the item for the requested epoch.
-	item, ok := lru.cache.Get(epoch)
+	item, ok := lru.cache.Get(cacheKey)
 	if !ok {
 		if lru.future > 0 && lru.future == epoch {
 			item = lru.futureItem
 		} else {
 			log.Trace("Requiring new ethash "+lru.what, "epoch", epoch)
 			item = lru.new(epoch, epochLength)
 		}
-		lru.cache.Add(epoch, item)
+		lru.cache.Add(cacheKey, item)
 	}
 
 	// Ensure pre-generation handles ecip-1099 changeover correctly
 	var nextEpoch = epoch + 1
 	var nextEpochLength = epochLength
 	if ecip1099FBlock != nil {
 		nextEpochBlock := nextEpoch * epochLength
-		if nextEpochBlock == *ecip1099FBlock && epochLength == epochLengthDefault {
+		if nextEpochBlock >= *ecip1099FBlock && epochLength == epochLengthDefault {
```

Comment:
Introduced at commit f4c42dcc40b7b2778957944141d2edd1d972f038.
I think maybe I'd better revert this; it may be too generous.

Suppose ECIP1099 is activated at block 3_000_042.
Then the `nextEpoch` will begin at 3_030_000, lasting till 3_090_000. 

From the [spec](https://github.com/ethereumclassic/ECIPs/blob/master/_specs/ecip-1099.md#implementation):
> For the smoothest possible transition activation should occur on a block in which an epoch transition to an even epoch number is occurring.

Agree?

Reply:
Yes, I would stick to the spec. Haven't checked the spec while reviewing.

Reply:
Resolved with https://github.com/etclabscore/core-geth/pull/499/commits/545a964e57fa490dcd8aa973e5a2ecc268625fa7